docs/dgram.md

```diff
@@ -0,0 +1,98 @@
+ZJS API for UDP datagram sockets
+================================
+
+* [Introduction](#introduction)
+* [Web IDL](#web-idl)
+* [API Documentation](#api-documentation)
+* [Client Requirements](#requirements)
+* [Sample Apps](#sample-apps)
+
+Introduction
+------------
+The `dgram` API is based on the subset of the
+[corresponding module](https://nodejs.org/api/dgram.html) in Node.js.
+It allows to send and receive UDP datagrams.
+
+Web IDL
+-------
+This IDL provides an overview of the interface; see below for documentation of
+specific API functions.
+
+```javascript
+// require returns a socket factory object
+// var dgram = require('dgram');
+
+[NoInterfaceObject]
+interface dgram {
+    DgramSocket createSocket(string udp4_or_udp6);
+};
+
+[NoInterfaceObject]
+interface DgramSocket {
+    void on(string event, RecvCallback cb);
+    void bind(int port, string ip_addr);
+    void send(Buffer buf, unsigned long offset, unsigned long len, int port, string ip_addr, [SendCallback cb]);
+    void close();
+};
+
+callback RecvCallback = void (Buffer msg, RemoteInfo rinfo);
+callback SendCallback = void (Error err);  // or undefined if no error
+
+
+callback EventCallback = void (various);  // callback arg depends on event
+
+dictionary RemoteInfo {
+    string ip_addr;
+    string family;
+    int port;
+};
+```
+
+API Documentation
+-----------------
+### dgram.createSocket
+
+`DgramSocket createSocket(string type);`
+
+Create a datagram socket of given type, which must be `'udp4'` or `'udp6'`.
+
+### DgramSocket.on
+
+`void on(string event, RecvCallback callback);`
+
```

Comment:
Regarding state of Z net stack: the biggest issue with it (now that it basically works) exactly that many error conditions aren't detected. Instead, it can just hang (like was due to lack of net buffers, but that's now fixed in Z version to which you guys updated Z.js recently, hurrah! I updated my code for it too), or just work in unexpected ways.

Beyond that, UDP is a simple thing and you can get only so much "errors" with it. Where you can get it, e.g. with send, I implemented error callback (which gets passed to the method). So, testing is the biggest issue here (plus some interactions which aren't obvious from Node.js docs, as we're really entering "callback hell" area with various callbacks in different places).

Anyway, if you suggest it to be done, I can go for that, test by injecting a fault by patching Z (that's what I did for send error callback), and report it as done ;-).